src/persist/src/indexed/columnar/parquet.rs

```diff
@@ -124,25 +137,57 @@ fn encode_parquet_kvtd<W: Write>(
 
 fn decode_parquet_file_kvtd<R: Read + Seek>(
     r: &mut R,
+    format_metadata: Option<ProtoFormatMetadata>,
     metrics: &ColumnarMetrics,
-) -> Result<Vec<ColumnarRecords>, Error> {
+) -> Result<BlobTraceUpdates, Error> {
     let metadata = read_metadata(r)?;
     let schema = infer_schema(&metadata)?;
-    let reader = FileReader::new(r, metadata.row_groups, schema, None, None, None);
+    let row_group_count = metadata.row_groups.len();
 
-    let file_schema = reader.schema().fields.as_slice();
-    // We're not trying to accept any sort of user created data, so be strict.
-    if file_schema != SCHEMA_ARROW_KVTD.fields {
-        return Err(format!(
-            "expected arrow schema {:?} got: {:?}",
-            SCHEMA_ARROW_KVTD.fields, file_schema
-        )
-        .into());
-    }
+    let mut reader = FileReader::new(r, metadata.row_groups, schema, None, None, None);
 
-    let mut ret = Vec::new();
-    for batch in reader {
-        ret.push(decode_arrow_batch_kvtd(&batch?, metrics)?);
+    let file_schema = reader.schema().fields.as_slice();
+    match format_metadata {
+        None => {
+            // We're not trying to accept any sort of user created data, so be strict.
+            if file_schema != SCHEMA_ARROW_KVTD.fields {
+                return Err(format!(
+                    "expected arrow schema {:?} got: {:?}",
+                    SCHEMA_ARROW_KVTD.fields, file_schema
+                )
+                .into());
+            }
+
+            let mut ret = Vec::new();
+            for batch in reader {
+                let records = decode_arrow_batch_kvtd(&batch?, metrics, 4)?;
+                ret.push(records);
+            }
+            Ok(BlobTraceUpdates::Row(ret))
+        }
+        Some(ProtoFormatMetadata::StructuredMigration(1)) => {
+            // It's an invariant that we only write a single Row Group.
```

Comment:
OOC, why change this as part of the migration? (If we've ever hashed this out together I've forgotten the details!)

Reply:
We didn't hash this out, and is something I need to add to the design doc!

Our current stats architecture (i.e. `PartStats`) represents an entire part. If we have more than one Row Group then we'd end up generating multiple `PartStats` that we would need to merge together. I made this change actually but when chatting with @danhhz we realized having a single Row Group is also nice for projection pushdown.

AFAIU [a single Parquet file is broken into Row Groups](https://parquet.apache.org/docs/concepts/), and each Row Group is broken into Column chunks. If we had a table with columns 'a', 'b', and 'c', and a parquet file with multiple Row Groups, then to fetch column 'a' we would need to issue multiple GET requests to S3 to get the 'a' Column Chunks in each Row Group. By enforcing a single Row Group it means we need only a single GET request to fetch a column.

I'm [pretty sure we already write a single Row Group](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/internal/encoding.rs#L1374-L1382) but formally enforcing it seems like a good idea. Also, switching from writing a single Row Group to multiple Row Groups seems much easier to do than the reverse!

Reply:
Understood, thanks!